### ISSUE TITLE
Catch exceptions from FreeMarker template files

### ISSUE THREAD

**Issue by danielsawan** [ninjaframework/ninja#171]
I already asked for this new fonctionality in google groups : 

https://groups.google.com/forum/?hl=fr#!topic/ninja-framework/gjYih9fjwiE

**Comment by raphaelbauer** [comment]
Hey @danielsawan ,

If you want to implement the feature. The key part is here:

https://github.com/ninjaframework/ninja/blob/develop/ninja-core/src/main/java/ninja/utils/ResultHandler.java

I'd create an interface that handles the errors + one default implementation. Then you can simply swap the implementation by specifying your own.

Something like that:
public interface Global {
  Result onError(Throwable t);
}

That's of course open to suggestions :)

**Comment by danielsawan** [comment]
i did a quick review of the code and i found that the "problem" is in NinjaImpl.java line 110 :

```
  try {

                Result result = route.getFilterChain().next(context);

                resultHandler.handleResult(result, context);

            } catch (Exception e) {

                logger.error(
                        "Emitting bad request 400. Something really wrong when calling route: {} (class: {} method: {})", 
                        context.getRequestPath(), route.getControllerClass(), route.getControllerMethod(), e);

                 Result result = Results.html().status(Result.SC_400_BAD_REQUEST).template(
                        NinjaConstant.LOCATION_VIEW_FTL_HTML_BAD_REQUEST);

                 resultHandler.handleResult(result, context);

            }
```

As you can see the template Exception is catched here and never throwed again. Instead we return an http 400 error without details.

**Comment by raphaelbauer** [comment]
Yes! You are absolutely right.

We should have that "Global" thing I wrote above here. inside the {catch (Exception e) {..} we can simply call global.onError(e) or so...

Does that make roughly make sense? Do you have any other ideas?

**Comment by danielsawan** [comment]
Oh sorry but i am definitely not an expert in ninja web framework, i am only training for 2 days. I do not know yet the architecture of the framework I am only focusing the missing needs and I think that a well-formed framework must catch all errors and give the possibility to dev to make things with all the non caught exception.The problem here is that exceptions threw by templates are just logged and that is not enough.

Same problem for the controller exception, the dev must not forget to make a global Filter in order to be able to catch exceptions from all controllers in one place (even if it is written in docs). Nevertheless, I think that this must be directly included in the framework.

Play global class in the way to go I think : http://www.playframework.com/documentation/2.0/ScalaGlobal

Unfortunately, I am not able to do it myself because in am very new to this framework and have no much time.

**Comment by raphaelbauer** [comment]
No worries. It is always good getting feedback about features people need and their priorities! Thanks!

**Comment by danielsawan** [comment]
Any news about this bug fix ? This is the only thing I am waiting to be fixed before starting my project on ninja :/

**Comment by raphaelbauer** [comment]
Hi Daniel,

Ninja is open source and lives from community contributions. People like you make it happen :) I don't think it will take much time to implement - even for non-experienced persons. If you want to make first steps let me know :) We can always have a hangout (timezone cest in the evening) to discuss the implementation :)

If it's blocking you from starting a project with Ninja consider implementing the solution :) You'll quickly discover that it's easy to extend Ninja - which in turn gives you even more confidence in Ninja :)

Cheers,

Raphael

**Comment by raphaelbauer** [comment]
Hi Daniel,

have a look at develop. It now supports everything you need and a lot more.

Cheers,

Raphael

**Comment by danielsawan** [comment]
oh thx ! i can't check it now coz i am on vacation in USA will test this when i am back to France. Thanks again !

**Comment by danielsawan** [comment]
Hi i tested the new way of managing error en 4.0.4 but it seems that the framework still don't catch error coming from templates.

I simply put an invalid variable on my .ftl.html file like ${rhth} and refresh the browser.

I also had override all the error method in order to see if the error is catched. Here are all the methods in conf.Ninja

```
@Override
public Result onException(Context context, Exception exception) {
    return super.onException(context, exception);
}

@Override
public Result getBadRequestResult(Context context, Exception exception) {
    return super.getBadRequestResult(context, exception);
}

@Override
public Result getInternalServerErrorResult(Context context,
        Exception exception) {
    return super.getInternalServerErrorResult(context, exception);
}

@Override
public void renderErrorResultAndCatchAndLogExceptions(Result result,
        Context context) {
    super.renderErrorResultAndCatchAndLogExceptions(result, context);
}    

@Override
public void onRouteRequest(Impl context) {
        super.onRouteRequest(context);  
}

@Override
public Result getForbiddenResult(Context context) {
    return super.getForbiddenResult(context);
}

@Override
public Result getNotFoundResult(Context context) {
    // TODO Auto-generated method stub
    return super.getNotFoundResult(context);
}

@Override
public Result getUnauthorizedResult(Context context) {
    // TODO Auto-generated method stub
    return super.getUnauthorizedResult(context);
}
```

None of all those method seems to be called when the template error occured.

**Comment by danielsawan** [comment]
Oh i just noticed that when i code an aritmetic bug like {1 / 0} the error is catched as it should. Seems like syntax error are not catch, not a big deal but still...

**Comment by raphaelbauer** [comment]
Can you try in production mode as well? I think the behaviour you are describing is expected in dev.

**Comment by danielsawan** [comment]
I just try in prod to put a wrong variable in template like ${test123}. It throw an error on log which is : freemarker.core.InvalidReferenceException: The following has evaluated to null or missing

And show on the browser : http://puu.sh/eUiOl/8f13f6826d.png

But nothing is intercepted on any method in the Ninja.java file

**Comment by raphaelbauer** [comment]
Good catch! Could you create a new ticket including the description and the screenshot? Thanks!
